### PR TITLE
Fix #121, fix #122: Improve the front-end API

### LINKED ISSUE #121: The "Show outliers" button is not working due to a regression

### LINKED ISSUE #122: The detail page is loading the wrong chart due to a regression

### PR DESCRIPTION
**Pull request by openjck**
This commit improves the front-end API. Each chart is now loaded
individually using Redux. The detail page and "Show outliers" button are
fixed in the process.

This setup more closely resembles the setup in Brad Westfall's example.
The store has a single `metrics` object (just like his store has a
single `users` object[1]) and metric data is added to it over time.
Components subscribe to only the metric data they care about.

As in Brad Westfall's example, download status is no longer considered
state. The `metrics` object isn't downloaded; it's assembled. It has no
download status of its own. Thankfully, the download status of
individual metrics can be inferred in container components by checking
for the existence of `this.props.metric`.

[1] https://github.com/bradwestfall/CSS-Tricks-React-Series/blob/376e1b0f37b2fdebabd465063495d7186a91d163/guide-3-redux/app/reducers/user-reducer.js

### CODE REVIEW

**Comment by openjck** [comment]
The diff looks scarier than it is. A lot of changes result from renaming things along the way.

File: distributionviewer/core/static/js/app/components/containers/chart-container.js [hunk 1]
```diff
@@ -1,10 +1,39 @@
...
+          {...this.props}
+        />
+      );
+    }
+  }
 }
+
+const mapStateToProps = function(store, ownProps) {
+  return {
+    metric: store.metricState.metrics[ownProps.metricId],
```
Comment (spasovski) [inline distributionviewer/core/static/js/app/components/containers/chart-container.js:35]: Great find! This is what we were looking for all this time.

**Comment by spasovski** [comment]
I am not a fan of the name "MetricMetadata" since to me it's a metric list (granted it has metadata) but I wouldn't block on it. I will however ask that you do this to the origin d3v4 branch instead of master since we need to merge that pronto and move forward with missing features.

**Comment by openjck** [comment]
> I am not a fan of the name "MetricMetadata" since to me it's a metric list (granted it has metadata) but I wouldn't block on it. I will however ask that you do this to the origin d3v4 branch instead of master since we need to merge that pronto and move forward with missing features.

Agreed. Names matter, and I'd be happy to go with another one. `metricList` sounds good, but I wouldn't know what to call a single member of that object. Not `metric` since that would be confusing. I'll keep thinking about it.

**Pull request closed without merging**